core/modules/css/KvInterfaceImplMySql.cc

```diff
@@ -0,0 +1,503 @@
+    sql::SqlTransaction _trans;
+};
+
+
+KvInterfaceImplMySql::KvInterfaceImplMySql(mysql::MySqlConfig const& mysqlConf)
+: _conn(mysqlConf) {
+
+}
+
+
+void
+KvInterfaceImplMySql::_findParentId(std::string const& childKvKey, bool* hasParent, unsigned int* parentKvId) {
+
+    // This should only be called when a transaction is in process.
+
+    if (hasParent == NULL || parentKvId == NULL) {
+        throw CssError("null out-var ptr");
+    }
+
+    size_t loc = childKvKey.find_last_of(KEY_PATH_DELIMITER);
+    if (loc == 0) {
+        *hasParent = false;
+        return;
+    }
+    std::string parentKey = childKvKey.substr(0, loc);
+    boost::format fmQuery("SELECT kvId FROM kvData WHERE kvKey='%1%'");
+    fmQuery % parentKey;
+    std::string query = fmQuery.str();
+    sql::SqlResults results;
+    sql::SqlErrorObject errObj;
+    if (not _conn.runQuery(query, results, errObj)) {
+        LOGF(_logger, LOG_LVL_ERROR, "_findParentId - query failed: %1%" % query);
+        throw CssError(errObj);
+    } else {
+        // If the key is not found then the parent does not exist; create it.
+        if (not extractIntValueFromSqlResults(&results, parentKvId)) {
+            // _create will throw if parent can't be created.
+            *parentKvId = _create(parentKey, "");
+        }
+        *hasParent = true;
+    }
+}
+
+
+void
+KvInterfaceImplMySql::create(std::string const& key, std::string const& value) {
+    sql::SqlErrorObject errObj;
+    std::string vKey = _validateKey(key);
+    std::string vValue = _validateValue(value);
+    KvTransaction transaction(_conn);
+    _create(vKey, vValue);
+    transaction.commit();
+}
+
+
+unsigned int
+KvInterfaceImplMySql::_create(std::string const& key, std::string const& value) {
+    // A transaction should be in process here.
+
+    unsigned int parentKvId(0);
+    bool hasParent(false);
+    _findParentId(key, &hasParent, &parentKvId);
+
+    boost::format fmQuery;
+    if (hasParent) {
+        fmQuery = boost::format("INSERT INTO kvData (kvKey, kvVal, parentKvId) VALUES ('%1%', '%2%', '%3%')");
+        fmQuery % key % value % parentKvId;
+    } else {
+        fmQuery = boost::format("INSERT INTO kvData (kvKey, kvVal) VALUES ('%1%', '%2%')"); // leave parentKvId NULL
+        fmQuery % key % value;
+    }
+    std::string query = fmQuery.str();
+
+    // run query
+    sql::SqlErrorObject errObj;
+    if (not _conn.runQuery(query, errObj)) {
+        switch (errObj.errNo()) {
+        default:
+            throw CssError(errObj);
+            break;
+
+        case ER_DUP_ENTRY:
+            LOGF(_logger, LOG_LVL_ERROR, "_create - SQL INSERT INTO failed: %1%" % query);
+            throw KeyExistsError(errObj);
+            break;
+        }
+    }
+
+    std::string getIdQuery = "SELECT LAST_INSERT_ID()";
+    sql::SqlResults results;
+    errObj.reset();
+    if (not _conn.runQuery(getIdQuery, results, errObj)) {
+        throw CssError(errObj);
+    }
```

Comment:
oh, I replaced those with _conn.getInsertId() when I found that method, but forgot to delete these. thanks.